src/OpenTelemetry.Instrumentation.Process/ProcessMetrics.cs

```diff
@@ -16,46 +16,76 @@
 
 using System.Diagnostics.Metrics;
 using System.Reflection;
+using System.Threading;
 using Diagnostics = System.Diagnostics;
 
 namespace OpenTelemetry.Instrumentation.Process;
 
-internal class ProcessMetrics
+internal sealed class ProcessMetrics
 {
     internal static readonly AssemblyName AssemblyName = typeof(ProcessMetrics).Assembly.GetName();
     internal static readonly Meter MeterInstance = new(AssemblyName.Name, AssemblyName.Version.ToString());
-    private static readonly Diagnostics.Process CurrentProcess = Diagnostics.Process.GetCurrentProcess();
 
-    static ProcessMetrics()
+    private static readonly ThreadLocal<InstrumentsValues> CurrentThreadInstrumentsValues = new(() => new InstrumentsValues());
+
+    public ProcessMetrics(ProcessInstrumentationOptions options)
     {
         // TODO: change to ObservableUpDownCounter
         MeterInstance.CreateObservableGauge(
             "process.memory.usage",
-            () =>
-            {
-                CurrentProcess.Refresh();
-                return CurrentProcess.WorkingSet64;
-            },
+            () => CurrentThreadInstrumentsValues.Value.GetMemoryUsage(),
             unit: "By",
             description: "The amount of physical memory in use.");
 
         // TODO: change to ObservableUpDownCounter
         MeterInstance.CreateObservableGauge(
             "process.memory.virtual",
-            () =>
-            {
-                CurrentProcess.Refresh();
-                return CurrentProcess.VirtualMemorySize64;
-            },
+            () => CurrentThreadInstrumentsValues.Value.GetVirtualMemoryUsage(),
             unit: "By",
             description: "The amount of committed virtual memory.");
     }
 
-    /// <summary>
-    /// Initializes a new instance of the <see cref="ProcessMetrics"/> class.
-    /// </summary>
-    /// <param name="options">The options to define the metrics.</param>
-    public ProcessMetrics(ProcessInstrumentationOptions options)
+    private sealed class InstrumentsValues
     {
+        private static readonly Diagnostics.Process CurrentProcess = Diagnostics.Process.GetCurrentProcess();
+        private double? memoryUsage;
+        private double? virtualMemoryUsage;
+
+        internal InstrumentsValues()
+        {
+            this.memoryUsage = null;
+            this.virtualMemoryUsage = null;
+        }
+
+        internal double GetMemoryUsage()
+        {
+            if (!this.memoryUsage.HasValue)
```

Comment:
nit: is it a code style to always use `this`? it is just more reading. In .NET repos we just define the fields prefixed with `_` and not using `this` ate all. it is easier to read.

Reply:
Yes. This is coming from the StyleCop Analyzer: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1101.md

> The advantage of using 'this.' is that it applies equally to all element types including methods, properties, etc., and not just fields, making all calls to class members instantly recognizable, regardless of which editor is being used to view the code. Another advantage is that it creates a quick, recognizable differentiation between instance members and static members, which are not prefixed.

> A final advantage of using the 'this.' prefix is that typing this. will cause Visual Studio to show the IntelliSense popup, making it quick and easy for the developer to choose the class member to call.

Reply:
Yes, but not mandatory rule to apply :-) as I mentioned we don't do that in the .NET code itself :-) we have other styling we apply. I am not asking to change anything more than wondering.

Reply:
Yes, I was also not used to using `this.` prefix earlier 😄

Reply:
I used to write `_foo` or `m_bar` before was treated as a C++ nerd who brings bad habits to .NET 🤣

Reply:
In runtime code we use `_foo` style, `s_bar` for static. This is actually the recommended style by us :-)

Reply:
You can tell that I was coming from DevDiv https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/215bc921a233200e2acfafc3420e1b763b4ea385/src/OpenTelemetry.Exporter.Geneva/GenevaExporterOptions.cs#L31 🤣

Reply:
To my knowledge, these came from https://en.wikipedia.org/wiki/Hungarian_notation, looks like .NET decided to replace `m_` with `_`.

![image](https://user-images.githubusercontent.com/17327289/193139935-e7f2dce4-50bb-424b-b4a0-b5a093492070.png)

Reply:
could be but Hungarian notation was listing the variable type in the field name too. like `arr`, 'l', 'i'...etc.